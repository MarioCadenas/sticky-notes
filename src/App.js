import React, { Component } from 'react';
import Note from './components/Note';
import Form from './components/Form';
import styled, { createGlobalStyle } from 'styled-components';

const GlobalReset = createGlobalStyle`
  body {
    margin: 0;
    padding: 0;
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Roboto", "Oxygen",
      "Ubuntu", "Cantarell", "Fira Sans", "Droid Sans", "Helvetica Neue",
      sans-serif;
    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
  }
`;
const AppContainer = styled.div`
  text-align: center;
`
const Notes = styled.div`
  display: inline-block;
  height: auto;
  width: 100%;
`;

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      notes: []
    }
  }

  componentDidMount() {
    this.setState({ notes: this.loadStorage() });
  }

  loadStorage = () => JSON.parse(localStorage.getItem('notes')) || [];

  updateStorage = (notes) => localStorage.setItem('notes', JSON.stringify(notes, null, 2));

  addNote = (note) => {
    this.setState(({ notes: oldNotes }) => {
      const notes = [...oldNotes, note];
      this.updateStorage(notes);
      return { notes };
    });
  }

  removeNote = (noteId) => {
    this.setState(({ notes: oldNotes }) => {
      const notes = oldNotes.filter((note, i) => i !== noteId);
      this.updateStorage(notes);
      return { notes };
    });
  }

  render() {
    const { addNote, removeNote, state: { notes } } = this;

    return (
      <AppContainer>
        <GlobalReset />
        <Notes>
          {
            notes.length ? (
              notes.map((note, index) => <Note key={index} removeNote={removeNote} {...note} id={index} />)
            ) : (
              <p>No notes to display.</p>
            )
          }
        </Notes>
        <Form addNote={addNote} />
      </AppContainer>
    );
  }
}

export default App;
