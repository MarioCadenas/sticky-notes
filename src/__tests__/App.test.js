import React from 'react';
import { shallow } from 'enzyme';
import App from '../App';

window.localStorage = (function () {
  let store = {};

  return {
    getItem(key) {
      return store[key] || null;
    },
    setItem(key, value) {
      store[key] = value.toString();
    },
    clear() {
      store = {}
    }
  };
})();

describe('App component', () => {
  describe('App with notes in localStorage', () => {
    it('should correctly render with notes', () => {
      const note = { text: 'test' };
      window.localStorage.setItem('notes', JSON.stringify([note]));
      const wrapper = shallow(<App />);
      expect(wrapper).toMatchSnapshot();
    });
    it('should correctly load notes from localStorage', () => {
      const note = { text: 'test' };
      window.localStorage.setItem('notes', JSON.stringify([note]));
      const wrapper = shallow(<App />);
      expect(wrapper.state('notes')).toEqual([note]);
    });
    it('should not render message about empty notes', () => {
      const wrapper = shallow(<App />);
      expect(wrapper.find('p').length).toEqual(0);
    });
  });
  describe('App with empty notes', () => {
    beforeAll(() => {
      localStorage.clear();
    });
    it('should correctly render with empty notes', () => {
      const wrapper = shallow(<App />);
      expect(wrapper).toMatchSnapshot();
    });
    it('should have empty notes array if localStorage is empty', () => {
      const wrapper = shallow(<App />);
      expect(wrapper.state('notes')).toEqual([]);
    });
    it('should render message about empty notes', () => {
      const wrapper = shallow(<App />);
      expect(wrapper.find('p').length).toEqual(1);
      expect(wrapper.find('p').text()).toBe('No notes to display.');
    });
  });
});