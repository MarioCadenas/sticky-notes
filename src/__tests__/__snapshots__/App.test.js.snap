// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component App with empty notes should correctly render with empty notes 1`] = `
<styled.div>
  <GlobalStyleComponent />
  <styled.div>
    <p>
      No notes to display.
    </p>
  </styled.div>
  <Form
    addNote={[Function]}
  />
</styled.div>
`;

exports[`App component App with notes in localStorage should correctly render with notes 1`] = `
<styled.div>
  <GlobalStyleComponent />
  <styled.div>
    <Note
      id={0}
      key="0"
      removeNote={[Function]}
      text="test"
    />
  </styled.div>
  <Form
    addNote={[Function]}
  />
</styled.div>
`;
